examples/pokemon-service/src/authz.rs

```diff
@@ -0,0 +1,176 @@
+// the inner service's types, and _at most_ require that those be `Op::Input` and `Op::Error`, for
+// maximum composability:
+//
+// ```
+// ...
+// where
+//     S: Service<(Op::Input, ($($var,)*)), Error = Op::Error>
+//     ...
+// {
+//     type Response = S::Response;
+//     type Error = S::Error;
+//     type Future = Pin<Box<dyn Future<Output = Result<S::Response, S::Error>> + Send>>;
+// }
+//
+// ```
+//
+// This plugin still exemplifies how changing a type can be done to make it more interesting.
+
+impl<P, E> IntoResponse<P> for AuthorizeServiceError<E>
+where
+    E: IntoResponse<P>,
+{
+    fn into_response(self) -> http::Response<BoxBody> {
+        match self {
+            AuthorizeServiceError::InnerServiceError(e) => e.into_response(),
+            AuthorizeServiceError::AuthorizeError { message } => http::Response::builder()
+                .status(http::StatusCode::UNAUTHORIZED)
+                .body(aws_smithy_http_server::body::to_boxed(message))
+                .expect("attempted to build an invalid HTTP response; please file a bug report"),
+        }
+    }
+}
+
+macro_rules! impl_service {
+    ($($var:ident),*) => {
+        impl<S, Op, $($var,)*> Service<(Op::Input, ($($var,)*))> for AuthorizeService<Op, S>
+        where
+            Op: OperationShape,
+            S: Service<(Op::Input, ($($var,)*)), Error = Op::Error> + Clone + Send + 'static,
+            S::Future: Send,
+            Op::Input: Send + 'static,
+            $($var: Send + 'static,)*
+        {
+            type Response = S::Response;
+            type Error = AuthorizeServiceError<Op::Error>;
+            type Future =
+                Pin<Box<dyn std::future::Future<Output = Result<S::Response, Self::Error>> + Send>>;
+
+            fn poll_ready(
+                &mut self,
+                cx: &mut std::task::Context<'_>,
+            ) -> std::task::Poll<Result<(), Self::Error>> {
+                self.inner
+                    .poll_ready(cx)
+                    .map_err(|e| Self::Error::InnerServiceError(e))
+            }
+
+            fn call(&mut self, req: (Op::Input, ($($var,)*))) -> Self::Future {
+                let (input, exts) = req;
+
+                // Replacing the service is necessary to avoid readiness problems.
+                // https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services
+                let service = self.inner.clone();
+                let mut service = std::mem::replace(&mut self.inner, service);
+
+                let fut = async move {
+                    // You could imagine that here we would make a more complex async call to perform the
+                    // actual authorization.
+                    // We would likely need to add bounds on `Op::Input`, `Op::Error`, or on data
+                    // types the plugin would hold if we wanted to do anything useful.
+                    let is_authorized = true;
```

Comment:
Nit: It would be beneficial to showcase a separate type specifically for handling authorization. For the time being, we could simply pass `Op::ID.name()` and `Headers` to this new type. Although this type could just return true for now, it would clearly demonstrate to users how to implement authorization using just those two parameters.

Also, if you annotate an operation with the `@auth` trait in the Smithy model, can you determine at runtime whether `Op` has this specific trait? For instance, does the Rust operation type implement a `AuthRequired` marker trait for all operations with this trait? If so, we should showcase that here.

Reply:
I added an `Authorizer` struct to showcase the former. This introduces some more bounds on the `Service` implementation for things to work out though.

> For instance, does the Rust operation type implement a AuthRequired marker trait for all operations with this trait? If so, we should showcase that here.

It doesn't, no, but that's the plan to implement AWS Auth.